### PR TITLE
Implement @cached

### PR DESCRIPTION
**Pull request by NullVoxPopuli**
Related to: https://github.com/glimmerjs/glimmer.js/pull/358

### CODE REVIEW

File: packages/@ember/-internals/metal/lib/cached.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
+// NOTE: copied from: https://github.com/glimmerjs/glimmer.js/pull/358
```
Comment (rwjblue) [inline packages/@ember/-internals/metal/lib/cached.ts:1]: Can probably tweak this comment to mention that both `glimmerjs/glimmer.js` and Ember have copies of this code so folks should make fixes/edits to both locations.

File: packages/@ember/-internals/metal/lib/cached.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+ * your app. Many getters and tracked properties are only accessed once, rendered,
+ * and then never rerendered, so adding `@cached` when it is unnecessary can
+ * negatively impact performance.
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const cached: PropertyDecorator = (...args: any[]) => {
+  const [target, key, descriptor] = args;
+
+  // Error on `@cached()`, `@cached(...args)`, and `@cached propName = value;`
+  if (DEBUG && target === undefined) throwCachedExtraneousParens();
```
Comment (rwjblue) [inline packages/@ember/-internals/metal/lib/cached.ts:67]: I think we need this to be feature flagged. Steps are basically:

1. Add a new entry (with a value of `null`) to https://github.com/emberjs/ember.js/blob/61be2e1edf820df1c7bbb245a5cbe02f4a875805/packages/%40ember/canary-features/index.ts#L14-L24
2. Add an entry here https://github.com/emberjs/ember.js/blob/61be2e1edf820df1c7bbb245a5cbe02f4a875805/packages/%40ember/canary-features/index.ts#L76
3. Update this code to use `import { YOUR_FEATURE_FLAG_NAME } from '@ember/canary-features';` and use `if (YOUR_FEATURE_FLAG_NAME) {` around the guts of this function.
4. Update the tests to be wrapped in the same sort of `if`

File: packages/@ember/-internals/metal/tests/cached/get_test.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      assert.verifySteps([], 'getter was not called again after repeated property access');
+
+      person.lastName = 'Larsen';
+      assert.verifySteps([], 'changing a property does not trigger an eager re-computation');
+
+      assert.strictEqual(person.fullName, 'Kenneth Larsen');
+      assert.verifySteps(['Kenneth Larsen'], 'accessing the property triggers a re-computation');
+    }
+
+    // https://github.com/ember-polyfills/ember-cached-decorator-polyfill/issues/7
```
Comment (rwjblue) [inline packages/@ember/-internals/metal/tests/cached/get_test.js:49]: lets kill this comment (it doesn't really provide much more context than the test itself already does)

**Review by rwjblue — COMMENTED** [review]
I left a few inline comments for things to do, but here are a couple more that didn't have an inline location:

* Can you use [Co-Authored-By](https://docs.github.com/en/github/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors) to give @buschtoens attribution?
* Can you update https://github.com/emberjs/ember.js/blob/61be2e1edf820df1c7bbb245a5cbe02f4a875805/packages/%40glimmer/tracking/index.ts#L1 re-export?

**Comment by NullVoxPopuli** [comment]
image

it worked! yay!

**Comment by buschtoens** [comment]
@NullVoxPopuli @rwjblue Thank you! 🙈❤️

File: packages/@ember/-internals/metal/lib/cached.ts (hunk 2 above)
Comment (NullVoxPopuli) [reply packages/@ember/-internals/metal/lib/cached.ts]: done

File: packages/@ember/-internals/metal/tests/cached/get_test.js (hunk 3 above)
Comment (NullVoxPopuli) [reply packages/@ember/-internals/metal/tests/cached/get_test.js]: done

File: packages/@ember/-internals/metal/lib/cached.ts (hunk 1 above)
Comment (NullVoxPopuli) [reply packages/@ember/-internals/metal/lib/cached.ts:1]: tweaked 🙃

**Review by rwjblue — APPROVED** [review]

**Pull request merged**